apps/api-extractor-model/src/items/ApiDeclaredItem.ts

```diff
@@ -42,13 +46,17 @@ export class ApiDeclaredItem extends ApiDocumentedItem {
     super.onDeserializeInto(options, context, jsonObject);
 
     options.excerptTokens = jsonObject.excerptTokens;
+    if (jsonObject.uid !== undefined) {
+      options.uid = DeclarationReference.parse(jsonObject.uid);
```

Comment:
@rbuckton Just curious, do we actually need to parse this?  Seems like the .api.json hierarchy already has all the information needed to reproduce this UID.  If so, we could think of this field as being serialized for informational purposes only.

Suppose the canonicalization rules are changed in the future.  If we deserialize the UID, we'll get the old canonical form, whereas if we generate it then we'd get the new canonical form.  I think the latter is more desirable, but I'm not 100% sure.

Reply:
I parse the UID for several reasons:
- Validation of the UID in the event some external tool performs preprocessing on the json.
- Handling a possible future case of the UID being customized by a doc comment.
- The UID was likely generated with full access to the TypeScript type checker. It's possible that some fidelity might be lost when constructing the UID merely from the object model.

That said, I don't have a strong preference.

Reply:
Thanks for explaining. If you can think of any discrepancies between the model and the type checker, I'd be interested to understand then and see if we can incorporate them into the model.

Reply:
BTW I realized that there is a subtlety to this "canonical" idea.  Suppose a package does this:

```ts
enum A { }
namespace A { }
export { A as B };
export { A as C };

// Let's say D is also an enum+namespace
export { D as E, D as F } from 'other-package';
```

How many "declarations" are there here to be canonicalized?  Today API Extractor's answer would be:

- `this-package!B:enum`
- `this-package!B:namespace`
- `this-package!C:enum`
- `this-package!C:namespace`
- `other-package!D:enum`
- `other-package!D:namespace`
- `this-package!E:import`  // modeled by AstImport today but not written to .api.json yet
- `this-package!F:import`

Unexported declarations like `A` are considered anonymous and the .d.ts rollups are free to fiddle with their names.  

And if an aliased symbol gets exported multiple times, each export is considered to be its own documentation object. `D` and `E` are aliases for the same symbol, but API Extractor sees them as not even being the same kind of thing.  This is justified because alias equivalence depends on the version of `other-package` that gets installed, which may be unknown (e.g. due to an NPM peer dependency).

The weirdest case is `B` and `C`, which today have all their members duplicated in the generated documentation.  Really one of them should be recast as `this-package!C:alias`, but without a special TSDoc tag, there's no deterministic way to choose.  (Fortunately this rarely arises in practice.)  

But even if we fixed that, I think I'd still consider `this-package!C:alias` to be "canonical" according to my definition, because when someone links that in the docs, they really are talking about `C` and not `B` or `A`.

I feel fairly content with API Extractor's simplifications, because philosophically I see TypeScript as two completely different languages:
- A rich type algebra that enables IntelliSense for every possible legacy JavaScript API
- The subset of Java-like "good parts" that's sufficient to build scalable modern code bases without everyone needing a PhD in type algebra, while ensuring ["one-- and preferably only one --obvious way"](https://www.python.org/dev/peps/pep-0020/) to write code

API Extractor is much more focused on the second one.  But I recognize that this "canonical name" idea may fall apart when applied to the compiler's world, where every file is a module, and namepaths can refer to unexported "inner types".  

Let me know if you have any thoughts about that.

Reply:
Hrmm... perhaps this:
```ts
- `this-package!E:import`  // modeled by AstImport today but not written to .api.json yet
- `this-package!F:import`
```

Should have been emitted like this:
```ts
- `this-package!E:enum`  // modeled by AstImport today but not written to .api.json yet
- `this-package!E:namespace`
- `this-package!F:enum`
- `this-package!F:namespace`
```

...so people can refer to individual declarations of an imported symbol.  But now I'm thinking I should just call it a **canonical reference** instead of a **canonical name**, because the thing being named is more nebulous than I original thought.

Reply:
Hmm... today these edge cases are absent from the .api.json.  For example [test-outputs/exportDuplicate](https://github.com/rbuckton/web-build-tools/blob/9d3f6fa22ee8c30f3ed7a10e8745319339818a39/build-tests/api-extractor-scenarios/etc/test-outputs/exportDuplicate/api-extractor-scenarios.api.json#L30) is missing most of the declarations from [api-extractor-scenarios.api.md](https://github.com/rbuckton/web-build-tools/blob/deeplyLinkedtypes/build-tests/api-extractor-scenarios/etc/test-outputs/exportDuplicate/api-extractor-scenarios.api.md).  We should get that fixed, since it's a super interesting case for your UID generation algorithm.

I opened https://github.com/microsoft/web-build-tools/issues/1404 to track this.

Reply:
I'm not sure the best way to handle this. From TypeScript's perspective, `B`, `C`, `E`, and `F` would all be an _alias_ for the real declaration. If it were me, I would have the canonical name for `B` and `C` be something like `this-package!B:alias` and `this-package!C:alias` respectively, with references to `this-package!A:enum`, etc. The problem being that DocFX doesn't understand this right now (and thus would require a custom template to support it).

The documentation might look something like:

---

## Enum `A` (local)
(id: `this-package!~A:enum` (local, non-exported))
(definition of `A` as an enum)

## Namespace `A` (local)
(id: `this-package!~A:namespace` (local, non-exported))
(definition of `A` as a namespace)

## Export `B`
(id: `this-package!B:alias`)
Export of:
- **Enum `A` (local)**
- **Namespace `A` (local)**

## Export `C`
(id: `this-package!C:alias`)
Export of:
- **Enum `A` (local)**
- **Namespace `A` (local)**

---

Alternatively, you define `B` and `C` per meaning:

---

## Enum `A` (local)
(id: `this-package!~A:enum` (local, non-exported))
(definition of `A` as an enum)

## Namespace `A` (local)
(id: `this-package!~A:namespace` (local, non-exported))
(definition of `A` as a namespace)

## Enum `B` (alias)
(id: `this-package!B:enum`)
Export for: **Enum `A` (local)**

## Namespace `B` (alias)
(id: `this-package!B:namespace`)
Export for: **Namespace `A` (local)**

## Enum `C` (alias)
(id: `this-package!C:enum`)
Export for: **Enum `A` (local)**

## Namespace `C` (alias)
(id: `this-package!C:namespace`)
Export for: **Namespace `A` (local)**

---

If you are doing .d.ts rollup, then you might just chose one of the two (`B` or `C`) to be the definition and just have the other one point to the first one.

For a re-export, you can either do the same as above or just use the IDs for the thing being exported.

Reply:
Another option would be to update DeclarationReference to support further disambiguation:

```
this-package!E:export(other-package!D:enum)
this-package!E:export(other-package!D:namespace)
```

Though that would obviously complicate the logic of resolving a partial ID like `E:enum` to the fully-qualified canonical version.

Reply:
> If it were me, I would have the canonical name for `B` and `C` be something like `this-package!B:alias` and `this-package!C:alias` respectively, with references to `this-package!A:enum`, etc.

Here's a more extreme example:

**package-1**
```ts
export function A(): void {
}

export interface A {
}
```

**package-2**
```ts
export { A as B } from 'package-1';
```

**package-3**
```ts
import { B } from 'package-2';

export class C implements B {
}
```

Now suppose that we want to make the hyperlink for `C`'s `implements B` excerpt.  In order to take us to the right DocFX page (and ultimately be able to follow hyperlinks to the right final definition), that reference needs to tell us that:

1\. The target is called `B`.
2\. The target comes from `package-2`
3\. The target is the interface, not the function

All 3 pieces of information are required for an accurate hyperlink.  Thus I'd argue the canonical reference must be `package-2!B:class`.

Now, generally your UID also tells us what kind of DocFX page we'll find when we follow the hyperlink.  Here, that requires an extra piece of information:

4\. The target is actually an alias for some other declaration.

But is this extra information important?  What would we need it for?  For a **declaration reference** my goal was merely to ensure that it uniquely identifies the target `ApiItem`.

When I click on the hyperlink for `C`'s `implements B` excerpt, I'd expect to go to a page something like this:


> ---
> 
> ## Interface `B` (alias)
> (id: `package-2!B:class`)
>
> Alias for: <a href="#">A</a> from <a href="#">package-1</a>
> 
> **Signature**
> ```ts
> export { A as B } from 'package-1';
> ```
> ---

Reply:
I agree, its somewhat similar to showing inherited members in a subclass's documentation.